File: CoreScopeViewer/Features/NodeDetail/NodeDetailScreen.swift
import SwiftUI

struct NodeDetailScreen: View {
    let node: MeshNode
    @Environment(AnalyzerSettings.self) private var settings
    @Environment(FavoritesStore.self) private var favoritesStore
    @State private var viewModel = NodeDetailViewModel()

    var body: some View {
        ScrollView {
            VStack(alignment: .leading, spacing: 16) {
                if viewModel.isLoading {
                    LoadingIndicator(title: "Loading node data…")
                        .frame(maxWidth: .infinity, alignment: .leading)
                }

                DataLoadStatusView(
                    lastUpdatedAt: viewModel.lastUpdatedAt,
                    errorMessage: viewModel.errorMessage,
                    hasContent: viewModel.health != nil || viewModel.paths != nil || viewModel.reach != nil,
                    retry: retryNodeData
                )

                NodeIdentityCard(node: node)

                if let reach = viewModel.reach {
                    NodeReachCard(reach: reach)
                } else if viewModel.isLoading {
                    NodeReachLoadingCard()
                }

                if let health = viewModel.health {
                    NodeHealthCard(stats: health.stats)
                    if !health.observers.isEmpty {
                        NodeObserversCard(observers: health.observers)
                    }
                }

                if let reach = viewModel.reach {
                    if !reach.links.isEmpty {
                        NodeLinksCard(links: reach.links)
                    }
                }

                if let paths = viewModel.paths, !paths.paths.isEmpty {
                    NodePathsCard(paths: paths.paths, total: paths.totalPaths)
                }
            }
            .padding(16)
            .padding(.bottom, 104)
            .adaptiveContentWidth()
        }
        .background(NodeScopeBackground())
        .navigationTitle(node.name ?? "Node")
        .navigationBarTitleDisplayMode(.inline)
        .toolbar {
            ToolbarItem(placement: .topBarTrailing) {
                favoriteButton
            }
        }
        .task {
            viewModel.configure(settings: settings)
            await viewModel.load(pubkey: node.publicKey)
        }
    }

    private var favoriteButton: some View {
        let isFavorite = favoritesStore.contains(
            kind: .node,
            entityID: node.publicKey,
            source: favoriteSource
        )
        return Button {
            favoritesStore.toggle(node: node, source: favoriteSource)
        } label: {
            Image(systemName: isFavorite ? "star.fill" : "star")
        }
        .tint(NodeScopeStyle.signal)
        .accessibilityLabel(isFavorite ? "Remove node from favorites" : "Add node to favorites")
    }

    private var favoriteSource: String {
        AnalyzerSettings.normalizedHost(settings.host)
    }

    private func retryNodeData() {
        Task {
            await viewModel.load(pubkey: node.publicKey)
        }
    }
}

private struct NodeReachLoadingCard: View {
    var body: some View {
        NodeDetailCard(title: "Mesh Reach", symbol: "network") {
            HStack(spacing: 8) {
                ForEach(0..<3, id: \.self) { _ in
                    VStack(alignment: .leading, spacing: 7) {
                        Capsule()
                            .fill(NodeScopeStyle.signal.opacity(0.16))
                            .frame(width: 34, height: 14)
                        Capsule()
                            .fill(Color.secondary.opacity(0.12))
                            .frame(height: 9)
                    }
                    .frame(maxWidth: .infinity, alignment: .leading)
                    .padding(10)
                    .background(NodeScopeStyle.signal.opacity(0.06), in: RoundedRectangle(cornerRadius: 12))
                }
            }

            HStack(spacing: 8) {
                MeshRouteActivityIndicator()
                Text("Measuring network reach")
                    .font(.caption)
                    .foregroundStyle(.secondary)
            }
        }
        .accessibilityElement(children: .combine)
        .accessibilityLabel("Loading mesh reach")
    }
}

private struct NodeIdentityCard: View {
    let node: MeshNode

    var body: some View {
        VStack(alignment: .leading, spacing: 14) {
            HStack(spacing: 12) {
                Image(systemName: NodeRoleStyle.symbolName(for: node.role))
                    .font(.title2.weight(.semibold))
                    .foregroundStyle(.white)
                    .frame(width: 50, height: 50)
                    .background(NodeRoleStyle.color(for: node.role), in: Circle())

                VStack(alignment: .leading, spacing: 3) {
                    Text(node.name ?? "Unnamed Node")
                        .font(.title2.bold())
                    HStack(spacing: 6) {
                        Text(node.role.capitalized)
                            .font(.caption.weight(.bold))
                            .foregroundStyle(NodeRoleStyle.color(for: node.role))
                        Text("· Seen \(RelativeTime.string(from: node.lastSeen))")
                            .font(.caption)
                            .foregroundStyle(.secondary)
                    }
                }
            }

            Divider().opacity(0.35)

            VStack(alignment: .leading, spacing: 5) {
                Text("Public Key")
                    .font(.caption2.weight(.semibold))
                    .foregroundStyle(.secondary)
                Text(node.publicKey.uppercased())
                    .font(.caption.monospaced())
                    .textSelection(.enabled)
                    .fixedSize(horizontal: false, vertical: true)
            }

            HStack(spacing: 18) {
                NodeDateMetric(label: "First Seen", date: node.firstSeen)
                NodeDateMetric(label: "Last Seen", date: node.lastSeen)
            }
        }
        .padding(16)
        .instrumentCard()
    }
}

private struct NodeDateMetric: View {
    let label: LocalizedStringKey
    let date: Date

    var body: some View {
        VStack(alignment: .leading, spacing: 2) {
            Text(label)
                .font(.caption2.weight(.semibold))
                .foregroundStyle(.secondary)
            Text(date, format: .dateTime.month().day().year())
                .font(.subheadline.weight(.semibold))
        }
        .frame(maxWidth: .infinity, alignment: .leading)
    }
}

private struct NodeHealthCard: View {
    let stats: NodeHealthStats

    var body: some View {
        NodeDetailCard(title: "Health", symbol: "waveform.path.ecg") {
            HStack(spacing: 8) {
                NodeMetricTile(value: stats.totalTransmissions.formatted(), label: "Transmissions")
                NodeMetricTile(value: stats.packetsToday.formatted(), label: "Today")
                NodeMetricTile(
                    value: stats.avgSnr.map { String(format: "%.1f", $0) } ?? "—",
                    label: "Avg SNR"
                )
            }

            HStack(spacing: 14) {
                Label("\(stats.totalObservations.formatted()) observations", systemImage: "ear.fill")
                if let avgHops = stats.avgHops {
                    Label("\(avgHops, specifier: "%.1f") avg hops", systemImage: "point.3.connected.trianglepath.dotted")
                }
            }
            .font(.caption.weight(.semibold))
            .foregroundStyle(.secondary)
        }
    }
}

private struct NodeMetricTile: View {
    let value: String
    let label: LocalizedStringKey

    var body: some View {
        VStack(alignment: .leading, spacing: 4) {
            Text(value)
                .font(.headline.monospacedDigit())
                .minimumScaleFactor(0.75)
            Text(label)
                .font(.caption2)
                .foregroundStyle(.secondary)
                .lineLimit(1)
        }
        .frame(maxWidth: .infinity, alignment: .leading)
        .padding(10)
        .background(NodeScopeStyle.signal.opacity(0.08), in: RoundedRectangle(cornerRadius: 12))
    }
}

private struct NodeObserversCard: View {
    let observers: [NodeObserverStat]
    @State private var isExpanded = false

    private var visibleObservers: ArraySlice<NodeObserverStat> {
        observers.prefix(isExpanded ? observers.count : 3)
    }

    var body: some View {
        NodeDetailCard(title: "Heard By", symbol: "ear.badge.waveform") {
            VStack(spacing: 0) {
                ForEach(visibleObservers) { observer in
                    HStack(spacing: 10) {
                        Circle()
                            .fill(NodeScopeStyle.healthy.opacity(0.14))
                            .frame(width: 32, height: 32)
                            .overlay {
                                Image(systemName: "antenna.radiowaves.left.and.right")
                                    .font(.caption)
                                    .foregroundStyle(NodeScopeStyle.healthy)
                            }
                        VStack(alignment: .leading, spacing: 2) {
                            Text(observer.observerName ?? observer.observerId)
                                .font(.subheadline.weight(.semibold))
                                .lineLimit(1)
                            HStack(spacing: 8) {
                                if let iata = observer.iata { Text(iata) }
                                if let avgSnr = observer.avgSnr { Text("\(avgSnr, specifier: "%.1f") dB") }
                            }
                            .font(.caption2)
                            .foregroundStyle(.secondary)
                        }
                        Spacer()
                        Text("\(observer.packetCount) pkts")
                            .font(.caption.weight(.semibold))
                            .foregroundStyle(NodeScopeStyle.signal)
                    }
                    .padding(.vertical, 9)
                    if observer.id != visibleObservers.last?.id {
                        Divider().opacity(0.3)
                    }
                }
            }
            if observers.count > 3 {
                ExpandSectionButton(isExpanded: $isExpanded, totalCount: observers.count, noun: "observers")
            }
        }
    }
}

private struct NodeReachCard: View {
    let reach: NodeReachResponse

    var body: some View {
        NodeDetailCard(title: "Mesh Reach", symbol: "network") {
            HStack(spacing: 8) {
                NodeMetricTile(value: reach.importance.neighborDegree.formatted(), label: "Neighbors")
                NodeMetricTile(value: reach.importance.bidirectionalLinks.formatted(), label: "Two-way Links")
                NodeMetricTile(value: reach.importance.directObservers.formatted(), label: "Observers")
            }
            Text("Measured across a \(reach.window.days)-day network window")
                .font(.caption)
                .foregroundStyle(.secondary)
        }
    }
}

private struct NodeLinksCard: View {
    let links: [ReachLink]
    @State private var isExpanded = false

    private var visibleLinks: ArraySlice<ReachLink> {
        links.prefix(isExpanded ? links.count : 3)
    }

    var body: some View {
        NodeDetailCard(title: "Links", symbol: "link") {
            VStack(spacing: 0) {
                ForEach(visibleLinks) { link in
                    HStack(spacing: 10) {
                        Image(systemName: link.bidir ? "arrow.left.arrow.right.circle.fill" : "arrow.right.circle.fill")
                            .foregroundStyle(link.bidir ? NodeScopeStyle.healthy : NodeScopeStyle.activity)
                        VStack(alignment: .leading, spacing: 2) {
                            Text(link.name)
                                .font(.subheadline.weight(.semibold))
                                .lineLimit(2)
                                .layoutPriority(1)
                            Text(link.bidir ? "Bidirectional" : "One way")
                                .font(.caption2)
                                .foregroundStyle(.secondary)
                        }
                        Spacer()
                        if let distanceKm = link.distanceKm {
                            Text("\(distanceKm, specifier: "%.1f") km")
                                .font(.caption.weight(.semibold))
                                .foregroundStyle(.secondary)
                        }
                    }
                    .padding(.vertical, 9)
                    if link.id != visibleLinks.last?.id { Divider().opacity(0.3) }
                }
            }
            if links.count > 3 {
                ExpandSectionButton(isExpanded: $isExpanded, totalCount: links.count, noun: "links")
            }
        }
    }
}

private struct NodePathsCard: View {
    let paths: [NodePath]
    let total: Int
    @State private var isExpanded = false

    private var visiblePaths: ArraySlice<NodePath> {
        paths.prefix(isExpanded ? paths.count : 3)
    }

    var body: some View {
        NodeDetailCard(title: "Known Paths", symbol: "point.topleft.down.curvedto.point.bottomright.up") {
            Text("\(total) resolved route\(total == 1 ? "" : "s")")
                .font(.caption)
                .foregroundStyle(.secondary)

            VStack(spacing: 0) {
                ForEach(visiblePaths) { path in
                    VStack(alignment: .leading, spacing: 6) {
                        Text(path.hops.map(\.name).joined(separator: " → "))
                            .font(.subheadline.weight(.medium))
                            .fixedSize(horizontal: false, vertical: true)
                        HStack(spacing: 10) {
                            Label("\(path.hops.count) hops", systemImage: "point.3.connected.trianglepath.dotted")
                            Text("Seen \(path.count)×")
                        }
                        .font(.caption2.weight(.semibold))
                        .foregroundStyle(NodeScopeStyle.signal)
                    }
                    .padding(.vertical, 10)
                    if path.id != visiblePaths.last?.id { Divider().opacity(0.3) }
                }
            }
            if paths.count > 3 {
                ExpandSectionButton(isExpanded: $isExpanded, totalCount: paths.count, noun: "paths")
            }
        }
    }
}

private struct ExpandSectionButton: View {
    @Binding var isExpanded: Bool
    let totalCount: Int
    let noun: String

    var body: some View {
        Button {
            withAnimation(.easeInOut(duration: 0.22)) {
                isExpanded.toggle()
            }
        } label: {
            HStack {
                Text(isExpanded ? "Show less" : "Show all \(totalCount) \(noun)")
                Spacer()
                Image(systemName: isExpanded ? "chevron.up" : "chevron.down")
            }
            .font(.caption.weight(.bold))
            .foregroundStyle(NodeScopeStyle.signal)
            .padding(.horizontal, 12)
            .padding(.vertical, 9)
            .background(NodeScopeStyle.signal.opacity(0.09), in: Capsule())
        }
        .buttonStyle(.plain)
    }
}

private struct NodeDetailCard<Content: View>: View {
    let title: LocalizedStringKey
    let symbol: String
    @ViewBuilder let content: Content

    init(title: LocalizedStringKey, symbol: String, @ViewBuilder content: () -> Content) {
        self.title = title
        self.symbol = symbol
        self.content = content()
    }

    var body: some View {
        VStack(alignment: .leading, spacing: 14) {
            Label(title, systemImage: symbol)
                .font(.headline)
                .foregroundStyle(NodeScopeStyle.signal)
            content
        }
        .padding(16)
        .instrumentCard()
    }
}

File: android/app/src/main/java/org/nodescope/android/feature/nodes/NodeDetailScreen.kt
package org.nodescope.android.feature.nodes

import android.content.Intent
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.text.selection.SelectionContainer
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.outlined.CallMade
import androidx.compose.material.icons.outlined.*
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.runtime.saveable.rememberSaveable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import java.time.ZoneId
import java.time.format.DateTimeFormatter
import org.nodescope.android.core.design.*
import org.nodescope.android.core.model.*
import org.nodescope.android.core.network.Loaded

/** Role colors shared with the map markers. */
fun roleColor(role: String?): Color = when (role?.lowercase()) {
    "repeater" -> Color(0xFFFFAA44)
    "room" -> Color(0xFF299EFF)
    "companion" -> Color(0xFF45C99D)
    "sensor" -> Color(0xFFB18AFF)
    else -> Color(0xFF65DDB4)
}
fun roleIcon(role: String?): ImageVector = when (role?.lowercase()) {
    "repeater" -> Icons.Outlined.CellTower
    "room" -> Icons.Outlined.Forum
    "companion" -> Icons.Outlined.Smartphone
    "sensor" -> Icons.Outlined.Thermostat
    else -> Icons.Outlined.Router
}

/**
 * Node identity plus the iOS detail sections: mesh reach, health, observers that heard it,
 * links and known paths. [initial] comes from loaded map data or saved snapshots; the health
 * response supplies the node when it is not in the current region's results.
 */
@Composable
fun NodeDetailScreen(
    model: NodeDetailViewModel?, host: String, publicKey: String, initial: MeshNode?,
    favorite: Boolean, onFavorite: (MeshNode) -> Unit, onMap: (MeshNode) -> Unit,
    onObserver: (id: String, name: String) -> Unit, onNode: (String) -> Unit, onAnalytics: (() -> Unit)? = null,
    onViewed: (MeshNode) -> Unit = {},
) {
    // Without a network layer (previews/tests) only the identity card is shown.
    val health = model?.health?.state?.collectAsStateWithLifecycle()?.value ?: Loaded()
    val reach = model?.reach?.state?.collectAsStateWithLifecycle()?.value ?: Loaded()
    val paths = model?.paths?.state?.collectAsStateWithLifecycle()?.value ?: Loaded()
    val key = host to publicKey
    LaunchedEffect(model, key) { model?.load(host, publicKey) }
    val now = rememberNow()
    // Never show another node's (or analyzer's) sections while this one loads.
    val sections = listOf(health, reach, paths).map { if (it.key == key) it else Loaded(loading = model != null) }
    val healthValue = health.value?.takeIf { health.key == key }
    val reachValue = reach.value?.takeIf { reach.key == key }
    val pathsValue = paths.value?.takeIf { paths.key == key }
    val node = healthValue?.node ?: initial
    // Nodes opened from links may only be known once health loads; record them for Recently Viewed.
    LaunchedEffect(node?.publicKey) { node?.let(onViewed) }

    LazyColumn(Modifier.fillMaxSize(), contentPadding = PaddingValues(16.dp), verticalArrangement = Arrangement.spacedBy(12.dp)) {
        if (model != null) item {
            LoadStatus(combinedStatus(sections),
                hasContent = healthValue != null || reachValue != null || pathsValue != null, now) { model.load(host, publicKey, force = true) }
        }
        item {
            if (node != null) IdentityCard(node, publicKey, favorite, now, onFavorite, onMap)
            else if (sections[0].loading) LinearProgressIndicator(Modifier.fillMaxWidth())
            else Text("This node is not in the loaded results for $host.", style = MaterialTheme.typography.bodyMedium)
        }
        reachValue?.let { value -> item { ReachCard(value) } }
        healthValue?.stats?.let { stats -> item { HealthCard(stats) } }
        onAnalytics?.let { open -> item { AnalyticsLinkCard(open) } }
        healthValue?.observers?.takeIf { it.isNotEmpty() }?.let { observers ->
            item { ObserversCard(observers.sortedByDescending { it.packetCount }, onObserver) }
        }
        reachValue?.links?.takeIf { it.isNotEmpty() }?.let { links -> item { LinksCard(links, onNode) } }
        pathsValue?.takeIf { it.paths.isNotEmpty() }?.let { value -> item { PathsCard(value) } }
    }
}

@Composable
private fun IdentityCard(node: MeshNode, publicKey: String, favorite: Boolean, now: Long, onFavorite: (MeshNode) -> Unit, onMap: (MeshNode) -> Unit) {
    val copy = rememberCopyAction()
    val context = LocalContext.current
    var menuOpen by remember { mutableStateOf(false) }
    var identityExpanded by rememberSaveable(publicKey) { mutableStateOf(false) }
    DetailCard {
        Row(verticalAlignment = Alignment.CenterVertically, horizontalArrangement = Arrangement.spacedBy(8.dp)) {
            Box(Modifier.size(44.dp).background(roleColor(node.role), CircleShape), contentAlignment = Alignment.Center) {
                Icon(roleIcon(node.role), null, tint = Color(0xFF14243A))
            }
            Column(Modifier.weight(1f), verticalArrangement = Arrangement.spacedBy(3.dp)) {
                Text(node.name?.takeIf(String::isNotBlank) ?: "Unnamed node", style = MaterialTheme.typography.titleLarge)
                Text(node.role.replaceFirstChar { it.uppercase() }, style = MaterialTheme.typography.labelMedium,
                    color = MaterialTheme.colorScheme.onSurfaceVariant)
                Text("Seen ${relativeTime(parseInstant(node.lastSeen), now)}", style = MaterialTheme.typography.bodySmall,
                    color = MaterialTheme.colorScheme.onSurfaceVariant)
            }
            IconToggleButton(checked = favorite, onCheckedChange = { onFavorite(node) }) {
                Icon(if (favorite) Icons.Outlined.Star else Icons.Outlined.StarBorder,
                    if (favorite) "Remove favorite" else "Save node")
            }
            Box {
                IconButton(onClick = { menuOpen = true }) { Icon(Icons.Outlined.MoreVert, "Node actions") }
                DropdownMenu(expanded = menuOpen, onDismissRequest = { menuOpen = false }) {
                    DropdownMenuItem(text = { Text("Copy key") }, leadingIcon = { Icon(Icons.Outlined.ContentCopy, null) },
                        onClick = { menuOpen = false; copy("Public key", publicKey, false) })
                    DropdownMenuItem(text = { Text("Share") }, leadingIcon = { Icon(Icons.Outlined.Share, null) }, onClick = {
                        menuOpen = false
                        context.startActivity(Intent.createChooser(Intent(Intent.ACTION_SEND).setType("text/plain")
                            .putExtra(Intent.EXTRA_TEXT, NodeScopeLink.node(publicKey).url), "Share node link"))
                    })
                }
            }
        }
        if (node.coordinate != null) Button(onClick = { onMap(node) }) {
            Icon(Icons.Outlined.Map, null, Modifier.size(18.dp)); Spacer(Modifier.width(8.dp)); Text("Show on map")
        }
        TextButton(onClick = { identityExpanded = !identityExpanded }) {
            Text(if (identityExpanded) "Hide technical identity" else "Technical identity")
            Spacer(Modifier.width(4.dp))
            Icon(if (identityExpanded) Icons.Outlined.ExpandLess else Icons.Outlined.ExpandMore, null)
        }
        if (identityExpanded) {
            HorizontalDivider(color = MaterialTheme.colorScheme.outlineVariant)
            Text("Public key", style = MaterialTheme.typography.labelMedium, color = MaterialTheme.colorScheme.onSurfaceVariant)
            SelectionContainer { Text(publicKey.uppercase(), fontFamily = FontFamily.Monospace, style = MaterialTheme.typography.bodySmall) }
            Row(horizontalArrangement = Arrangement.spacedBy(16.dp)) {
                DateMetric("First seen", node.firstSeen, Modifier.weight(1f))
                DateMetric("Last seen", node.lastSeen, Modifier.weight(1f))
            }
        }
    }
}

@Composable
private fun AnalyticsLinkCard(onClick: () -> Unit) {
    Card(onClick = onClick, colors = CardDefaults.cardColors(containerColor = MaterialTheme.colorScheme.surface)) {
        Row(Modifier.fillMaxWidth().padding(16.dp), verticalAlignment = Alignment.CenterVertically, horizontalArrangement = Arrangement.spacedBy(14.dp)) {
            Box(Modifier.size(48.dp).background(MaterialTheme.colorScheme.primary.copy(alpha = 0.13f), CircleShape), contentAlignment = Alignment.Center) {
                Icon(Icons.Outlined.QueryStats, null, tint = MaterialTheme.colorScheme.primary)
            }
            Column(Modifier.weight(1f)) {
                Text("Node analytics", style = MaterialTheme.typography.titleMedium)
                Text("Activity, signal, coverage, and peer trends", style = MaterialTheme.typography.labelMedium,
                    color = MaterialTheme.colorScheme.onSurfaceVariant)
            }
            Icon(Icons.Outlined.ChevronRight, null, tint = MaterialTheme.colorScheme.onSurfaceVariant)
        }
    }
}

@Composable
private fun DateMetric(label: String, value: String?, modifier: Modifier) {
    Column(modifier.semantics(mergeDescendants = true) {}) {
        Text(label, style = MaterialTheme.typography.labelSmall, color = MaterialTheme.colorScheme.onSurfaceVariant)
        Text(parseInstant(value)?.let { DateTimeFormatter.ofPattern("MMM d, yyyy").withZone(ZoneId.systemDefault()).format(it) } ?: "Unknown",
            style = MaterialTheme.typography.titleSmall)
    }
}

@Composable
private fun ReachCard(reach: NodeReach) {
    DetailCard("Mesh reach", Icons.Outlined.Hub) {
        reach.importance?.let { importance ->
            Row(horizontalArrangement = Arrangement.spacedBy(8.dp)) {
                MetricTile(importance.neighborDegree.toString(), "Neighbors", Modifier.weight(1f))
                MetricTile(importance.bidirectionalLinks.toString(), "Two-way links", Modifier.weight(1f))
                MetricTile(importance.directObservers.toString(), "Observers", Modifier.weight(1f))
            }
        }
        reach.window?.let { Text("Measured across a ${it.days}-day network window", style = MaterialTheme.typography.bodySmall,
            color = MaterialTheme.colorScheme.onSurfaceVariant) }
    }
}

@Composable
private fun HealthCard(stats: NodeHealthStats) {
    DetailCard("Health", Icons.Outlined.MonitorHeart) {
        Row(horizontalArrangement = Arrangement.spacedBy(8.dp)) {
            MetricTile(stats.totalTransmissions?.let(::compactCount) ?: "—", "Transmissions", Modifier.weight(1f))
            MetricTile(stats.packetsToday?.let(::compactCount) ?: "—", "Today", Modifier.weight(1f))
            MetricTile(stats.avgSnr?.let { "%.1f".format(it) } ?: "—", "Avg SNR", Modifier.weight(1f))
        }
        Row(horizontalArrangement = Arrangement.spacedBy(14.dp)) {
            stats.totalObservations?.let { MetricChip("${compactCount(it)} observations", Icons.Outlined.Hearing) }
            stats.avgHops?.let { MetricChip("%.1f avg hops".format(it), Icons.Outlined.Route) }
        }
    }
}

@Composable
private fun ObserversCard(observers: List<NodeObserverStat>, onObserver: (String, String) -> Unit) {
    var expanded by rememberSaveable { mutableStateOf(false) }
    DetailCard("Heard by", Icons.Outlined.Sensors) {
        Column {
            observers.take(if (expanded) observers.size else 3).forEachIndexed { index, observer ->
                if (index > 0) HorizontalDivider(color = MaterialTheme.colorScheme.outlineVariant.copy(alpha = 0.5f))
                val name = observer.observerName?.takeIf(String::isNotBlank) ?: observer.observerId.take(12)
                Row(Modifier.fillMaxWidth().clickable { onObserver(observer.observerId, name) }.padding(vertical = 9.dp),
                    verticalAlignment = Alignment.CenterVertically, horizontalArrangement = Arrangement.spacedBy(10.dp)) {
                    Box(Modifier.size(32.dp).background(HealthyGreen.copy(alpha = 0.14f), CircleShape), contentAlignment = Alignment.Center) {
                        Icon(Icons.Outlined.Sensors, null, Modifier.size(16.dp), tint = HealthyGreen)
                    }
                    Column(Modifier.weight(1f)) {
                        Text(name, style = MaterialTheme.typography.titleSmall, maxLines = 1, overflow = TextOverflow.Ellipsis)
                        Text(listOfNotNull(observer.iata, observer.avgSnr?.let { "%.1f dB".format(it) }, observer.avgRssi?.let { "%.0f dBm".format(it) })
                            .joinToString(" · "), style = MaterialTheme.typography.labelSmall, color = MaterialTheme.colorScheme.onSurfaceVariant)
                    }
                    Text("${compactCount(observer.packetCount)} pkts", style = MaterialTheme.typography.labelMedium, color = MaterialTheme.colorScheme.primary)
                    Icon(Icons.Outlined.ChevronRight, null, Modifier.size(18.dp), tint = MaterialTheme.colorScheme.onSurfaceVariant)
                }
            }
        }
        if (observers.size > 3) ExpandButton(expanded, observers.size, "observers") { expanded = !expanded }
    }
}

@Composable
private fun LinksCard(links: List<ReachLink>, onNode: (String) -> Unit) {
    var expanded by rememberSaveable { mutableStateOf(false) }
    DetailCard("Links", Icons.Outlined.Link) {
        Column {
            links.take(if (expanded) links.size else 3).forEachIndexed { index, link ->
                if (index > 0) HorizontalDivider(color = MaterialTheme.colorScheme.outlineVariant.copy(alpha = 0.5f))
                Row(Modifier.fillMaxWidth().clickable { onNode(link.pubkey) }.padding(vertical = 9.dp),
                    verticalAlignment = Alignment.CenterVertically, horizontalArrangement = Arrangement.spacedBy(10.dp)) {
                    Icon(if (link.bidir) Icons.Outlined.SyncAlt else Icons.AutoMirrored.Outlined.CallMade,
                        if (link.bidir) "Two-way" else "One way", tint = if (link.bidir) HealthyGreen else ActivityAmber)
                    Column(Modifier.weight(1f)) {
                        Text(link.label, style = MaterialTheme.typography.titleSmall, maxLines = 2, overflow = TextOverflow.Ellipsis)
                        Text(if (link.bidir) "Bidirectional" else if (link.theyHear > 0) "Heard by this node's neighbor" else "Heard by this node",
                            style = MaterialTheme.typography.labelSmall, color = MaterialTheme.colorScheme.onSurfaceVariant)
                    }
                    link.distanceKm?.let { Text("%.1f km".format(it), style = MaterialTheme.typography.labelMedium, color = MaterialTheme.colorScheme.onSurfaceVariant) }
                    Icon(Icons.Outlined.ChevronRight, null, Modifier.size(18.dp), tint = MaterialTheme.colorScheme.onSurfaceVariant)
                }
            }
        }
        if (links.size > 3) ExpandButton(expanded, links.size, "links") { expanded = !expanded }
    }
}

@Composable
private fun PathsCard(value: NodePaths) {
    var expanded by rememberSaveable { mutableStateOf(false) }
    DetailCard("Known paths", Icons.Outlined.Route) {
        Text("${value.totalPaths} resolved route${if (value.totalPaths == 1) "" else "s"}", style = MaterialTheme.typography.bodySmall,
            color = MaterialTheme.colorScheme.onSurfaceVariant)
        Column {
            value.paths.take(if (expanded) value.paths.size else 3).forEachIndexed { index, path ->
                if (index > 0) HorizontalDivider(color = MaterialTheme.colorScheme.outlineVariant.copy(alpha = 0.5f))
                Column(Modifier.padding(vertical = 10.dp), verticalArrangement = Arrangement.spacedBy(6.dp)) {
                    Text(path.hops.joinToString(" → ") { it.label }, style = MaterialTheme.typography.bodyMedium)
                    Row(horizontalArrangement = Arrangement.spacedBy(8.dp)) {
                        MetricChip("${path.hops.size} hops", Icons.Outlined.Route, MaterialTheme.colorScheme.primary)
                        MetricChip("Seen ${path.count}×", Icons.Outlined.Visibility, MaterialTheme.colorScheme.primary)
                    }
                }
            }
        }
        if (value.paths.size > 3) ExpandButton(expanded, value.paths.size, "paths") { expanded = !expanded }
    }
}

@Composable
private fun ExpandButton(expanded: Boolean, total: Int, noun: String, onClick: () -> Unit) {
    FilledTonalButton(onClick = onClick, modifier = Modifier.fillMaxWidth()) {
        Text(if (expanded) "Show less" else "Show all $total $noun", Modifier.weight(1f))
        Icon(if (expanded) Icons.Outlined.ExpandLess else Icons.Outlined.ExpandMore, null)
    }
}

@Composable
private fun MetricTile(value: String, label: String, modifier: Modifier) {
    Surface(modifier.semantics(mergeDescendants = true) {}, color = MaterialTheme.colorScheme.primary.copy(alpha = 0.08f), shape = MaterialTheme.shapes.small) {
        Column(Modifier.padding(10.dp), verticalArrangement = Arrangement.spacedBy(4.dp)) {
            Text(value, style = MaterialTheme.typography.titleMedium)
            Text(label, style = MaterialTheme.typography.labelSmall, color = MaterialTheme.colorScheme.onSurfaceVariant)
        }
    }
}

@Composable
private fun DetailCard(title: String? = null, icon: ImageVector? = null, content: @Composable ColumnScope.() -> Unit) {
    Card(colors = CardDefaults.cardColors(containerColor = MaterialTheme.colorScheme.surface)) {
        Column(Modifier.fillMaxWidth().padding(16.dp), verticalArrangement = Arrangement.spacedBy(14.dp)) {
            if (title != null) Row(verticalAlignment = Alignment.CenterVertically, horizontalArrangement = Arrangement.spacedBy(8.dp)) {
                icon?.let { Icon(it, null, Modifier.size(18.dp), tint = MaterialTheme.colorScheme.primary) }
                Text(title, style = MaterialTheme.typography.titleMedium, color = MaterialTheme.colorScheme.primary)
            }
            content()
        }
    }
}
